Refactor variant report code and remove old functions

matren395 opened the pull request:
This PR will clean up the variant report code, by bringing it in-line with method standards and making the code in gks_compute_seqloc_digest_batch() less complex (changing it to read in native Python data - not in Hail - and to export native Python data)

- first step was removing the old functions and adding a fix requested by Phil Darnowsky

matren395 commented:
pending: big idea to reduce complexity is do our collect call much sooner , so that the dictionaries are only ever made in python . doesn't make sense to build them as hail objects if they are never used as hail objects

matren395 commented:
Push was only to the two functions in annotations.py to make the functions both accept and return python objects - changes to gnomad.py are about halfway done but coming well !

klaricch commented on gnomad/resources/grch38/gnomad.py:
```diff
@@ -593,46 +518,48 @@ def gnomad_gks_batch(
+                            input_dict=struct_list[xi],
+                            label_name="gnomAD",
+                            label_version=version,
+                            coverage_ht=coverage_ht,
```
add_gks_va_py is currently dependent on running add_gks_vrs_py because it requires the presence of "original_struct" can we drop the use of original_struct  throughout the functions, since all the information is already in struct_list?

klaricch commented on gnomad/resources/grch38/gnomad.py:
```diff
@@ -593,46 +518,48 @@ def gnomad_gks_batch(
+                            by_sex=by_sex,
+                            frequency_index=ht.freq_index_dict.collect()[0])
+
     outputs = []
```
can the use of outputs be dropped and just return out?

matren395 replied:
wouldn't returning 'out' only return one variant's dictionary instead of the list of multiple, which is the goal ?

klaricch replied:
oh i misread, nm!

matren395 commented on gnomad/resources/grch38/gnomad.py:
```diff
@@ -15,9 +15,8 @@
-    get_gks,
+    add_gks_va_py,
+    add_gks_vrs_py,
     gks_compute_seqloc_digest,
```
the function from gks_compute_seqloc_digest got coded into add_gks_vrs_py , so we definitely can remove this

currently only calls vrs_py and va_py from this

one question though - while working on them I added a _py suffix to differentiate them, since they're all in Python and not Hail anymore. Should I keep the suffix or is it unnecessary

matren395 commented on gnomad/resources/grch38/gnomad.py:
```diff
@@ -593,46 +518,48 @@ def gnomad_gks_batch(
-    if not vrs_only:
-        select_cols["gks_va_freq_json"] = hl.json(filtered.gks_va_freq_dict)
-    annotations = filtered.select(**select_cols).collect()  # might be big
+    # Filter and Select before adding annotations, so that everything can be kept in native python (is that what it's called)
```
nope actually, we can remove this (and in general clean up documentation on my ebnd)

matren395 commented on gnomad/resources/grch38/gnomad.py:
```diff
@@ -593,46 +518,48 @@ def gnomad_gks_batch(
+    # Add .vrs and .vrs_json (the JSON string representation of .vrs)
+    # Omits .location._id
+    
+    for xi in range(len(struct_list)):
```
Oh that is smart - definitely doable and a way to cut down on both time and complexity , I'll include it in the enxt commit

matren395 commented on gnomad/resources/grch38/gnomad.py:
```diff
@@ -593,46 +518,48 @@ def gnomad_gks_batch(
+    # Omits .location._id
+    
+    for xi in range(len(struct_list)):
+        struct_list[xi] = add_gks_vrs_py(struct_list[xi])
```
only hesitation is that it needs some information from struct_input.locus as well for the chromosome and reference genome - I think passing them in as separate arguments could work: 

add_gks_vrs_py(struct_locus=struct_list[xi].locus,struct_vrs=struct_list.info.vrs)

unless you'd prefer just passing in the build and chromosome in as their own arguments ?

klaricch replied:
i like the option of adding in the locus struct

matren395 commented on gnomad/utils/annotations.py:
```diff
@@ -1615,64 +1379,73 @@ def gks_compute_seqloc_digest_batch(
-    ht_out.vrs: Struct of the VRS representation of the variant
-    ht_out.vrs_json: JSON string representation of the .vrs struct.
+    :param struct_input: Struct input as the result of running .collect() on a HT
+    :return: Python Dictionary with fields vrs (Struct of the VRS representation of the variant)
```
that's doable - should be cleaner and the structure should be less confusing

matren395 commented on gnomad/utils/annotations.py:
```diff
@@ -1615,64 +1379,73 @@ def gks_compute_seqloc_digest_batch(
+    vrs_state_sequence = struct_input.info.vrs.VRS_States[1]
+
+    vrs_dict_out = {
+        "original_struct": struct_input,  # to remove later on
```
only pause was that the information from this would be needed for va_py() , but if I'm revising it to only return the vrs information then it's definitely possible to remove here

matren395 commented on gnomad/utils/annotations.py:
```diff
@@ -1681,6 +1454,7 @@ def add_gks_va(
     :param ancestry_groups_dict: Dict mapping shortened genetic ancestry group names to full names. Example: {'afr':'African/African American'} .
     :param by_sex: Boolean to include breakdown of ancestry groups by inferred sex (XX and XY) as well.
     :param vrs_only: Boolean to return only the VRS information and no general frequency information. Default is False.
+    :frequency_index: Index to map cohort to index in freq from ht.freq_index_dict[0] (make this nicer later)
```
I think I was waiting until a few more things were finalized (I/O for methods) before going back to the documentation, but I can definitely start working on it (especially for easy stuff like this) now

matren395 commented on gnomad/utils/annotations.py:
```diff
@@ -1778,36 +1548,38 @@ def _create_group_dicts(
+
+    if input_dict["original_struct"].popmax:
+        ancillaryResults = {
+            "homozygotes": overall_freq["homozygote_count"],
```
Note about this: it's not needed for homozygote count , but it is needed for "popMaxFAF95" below . I was encountering errors when running this as same variants' popmax field is just None , but I can be smarter about how this is implemented

matren395 commented on gnomad/utils/annotations.py:
```diff
@@ -1819,19 +1591,19 @@ def _create_group_dicts(
-        ancillaryResults = ancillaryResults.annotate(
-            meanDepth=coverage_ht[ht.locus].mean
-        )
+        ancillaryResults["meanDepth"] = coverage_ht[
+            input_dict["original_struct"].locus
+        ].mean
```
I believe it was working for me, but I can check again !

matren395 commented:
Note: the most recent push doesn't include any edits to documentation , only changes to some of the dictionary logic and some structural changes to how the looping was done. I've got a trainride tomorrow where I can hopefully clean up some of it!

matren395 commented:
All the commits above are from when I had to merge 'Main' into my own branch for this , so all the committs that would go into that are shown above

matren395 commented:
Just confirming: do we want to keep `by_ancestry_group` as an arg if the dictionary itself says `cohort` ? I feel like we confirmed that this is the case previously, I'm just double checking

matren395 commented:
Removed the redundant method , implemented it where it would be used , and tidied up some old/outdated docstrings. Should be merge ready, and definitely merge-ready with /variant-report-code

matren395 commented:
Link to new PR: https://github.com/broadinstitute/gnomad_methods/pull/597

klaricch commented:
Closing because copied to new PR https://github.com/broadinstitute/gnomad_methods/pull/597
